packages/react/checkbox/src/checkbox.test.tsx

```diff
@@ -20,184 +20,464 @@ global.ResizeObserver = class ResizeObserver {
+        render(
           <form
             onChange={(event) => {
               const target = event.target as HTMLInputElement;
-              expect(target.defaultChecked).toBe(false);
-              done(null);
+              onChange(target.checked);
             }}
           >
-            <CheckboxTest defaultChecked={false} />
+            <Checkbox.unstable_Provider defaultChecked>
+              <Checkbox.unstable_Trigger aria-label="basic checkbox">
+                <Checkbox.Indicator data-testid={INDICATOR_TEST_ID} />
+              </Checkbox.unstable_Trigger>
+              <Checkbox.unstable_BubbleInput />
+            </Checkbox.unstable_Provider>
           </form>
         );
-        fireEvent.click(checkbox);
-      }));
+
+        const checkbox = screen.getByRole(CHECKBOX_ROLE);
+        act(() => fireEvent.click(checkbox));
+        expect(onChange).toHaveBeenCalledWith(false);
+      });
+    });
+  });
+
+  describe('given a controlled Checkbox in form with bubble input', () => {
+    const onChange = vi.fn();
+
+    function ControlledCheckbox() {
+      const [checked, setChecked] = React.useState(false);
+      return (
+        <>
+          <Checkbox.unstable_Provider checked={checked} onCheckedChange={setChecked as any}>
+            <Checkbox.unstable_Trigger aria-label="basic checkbox">
+              <Checkbox.Indicator data-testid={INDICATOR_TEST_ID} />
+            </Checkbox.unstable_Trigger>
+            <Checkbox.unstable_BubbleInput />
+          </Checkbox.unstable_Provider>
+          <button type="button" onClick={() => setChecked((prev) => !prev)}>
+            Toggle checkbox
+          </button>
+        </>
+      );
+    }
+
+    it('should receive change event with controlled state when clicked', async () => {
+      render(
+        <form
+          onChange={(event) => {
+            const target = event.target as HTMLInputElement;
+            onChange(target.checked);
+          }}
+        >
+          <ControlledCheckbox />
+        </form>
+      );
+
+      const checkbox = screen.getByRole(CHECKBOX_ROLE);
+      act(() => fireEvent.click(checkbox));
+      expect(onChange).toHaveBeenCalledWith(true);
+
+      act(() => fireEvent.click(checkbox));
+      expect(onChange).toHaveBeenCalledWith(false);
+    });
+
+    it('should receive change event with controlled state when set externally', async () => {
+      render(
+        <form
+          onChange={(event) => {
+            const target = event.target as HTMLInputElement;
+            onChange(target.checked);
+          }}
+        >
+          <ControlledCheckbox />
+        </form>
+      );
+
+      const toggleButton = screen.getByText('Toggle checkbox');
+      act(() => fireEvent.click(toggleButton));
+      expect(onChange).toHaveBeenCalledWith(false);
+    });
   });
 });
 
-describe('given a controlled Checkbox in a form', () => {
-  afterEach(cleanup);
+describe('Legacy Checkbox', () => {
```

Comment:
Left these tests unchanged but moved them into a nested describe block, which is why the diff looks so big.